### ISSUE TITLE
Replication of the training process

### ISSUE THREAD

**Issue by GitHub-v7** [grammatical/neural-naacl2018#3]
Hello, I tried to re-trained the model twice following the readme file step by step but the M2 score was only 54.41 for CoNLL-14 test set. I found the recall is similar to the downloaded model but my re-trained model has a lower precision (either the ensemble or the single transformer model). Do you know which causes this issue? 

Besides, I notice you mentioned the pre-training embeddings and edit-weighted loss in your paper. but I have not seen any parameters related to these techniques in the transformer.sh script.  Where are these techniques used?

**Comment by snukky** [comment]
Hi,

You didn't mention any modifications of the training script, so I assume you trained the models on 4 GPUs with the workspace size predefined in the scripts. Could you provide the best validation M2 from one of your transformer models (from `models/transformer.?/valid.log`) and the best validation perplexity from the language model (from `models/lm.2/valid.log`)?

I think that if the language model is trained from scratch, it might be actually needed to tune its weight for the ensemble again. At the moment the weight is hardcoded in `train.sh` and set to 0.4. I grid searched it based on the CoNLL-2013 test set checking values between 0.1 and 1.0 with 0.1 interval, and picked one which gave the highest M2. This doesn't require re-training the models. I will add a script for this in spare time.

Transformer models are pre-trained with a language model (including embedding layers):

    --pretrained-model models/lm.1/model.npz.best-perplexity.npz \

The edit-weighted loss is applied via data weighting:

    --data-weighting-type word --data-weighting ./data/trainset.lc.bpe.cor.w3 \

Edit weights are calculated by the external script `tools/edit_weights.py`.

**Comment by GitHub-v7** [comment]
Hello,
Yes, I trained the models on 4 GPUs with the workspace size predefined in the scripts.

The best validation results of all models are shown below.

Re-trained model | cross-entropy | ce-mean-words | translation | perplexity
Transformer.1       | 13.3551           | 0.610185            | 0.4238       | 1.84077
Transformer.2       | 13.4716           | 0.615506            | 0.4219       | 1.85059
Transformer.3       | 13.3336           | 0.609201            | 0.4165       | 1.83896
Transformer.4       | 13.3155           | 0.608375            | 0.4204       | 1.83744
Lm.1                     |                         | 4.0617                |                   | 58.073
Lm.2                     |                         | 4.05344              |                   | 57.5953

I also tested all downloaded transformer models separately on the CoNLL-2014 test set and then compared them with the re-trained transformer models. The results are shown below.

Downloaded model   | Precision   | Recall   | M2
1                                 | 59.12         | 40.36   | 54.09
2                                 | 59.73         | 39.49   | 54.18
3                                 | 61.15         | 39.92   | 55.27
4                                 | 61.70         | 36.51   | 54.22

Re-trained model       | Precision   | Recall   | M2
1                                 | 56.18         | 40.37    | 52.10
2                                 | 56.81         | 40.96    | 52.73
3                                 | 56.42         | 39.11    | 51.83
4                                 | 56.52         | 40.02    | 52.21

The results are consistent with the ensemble, i.e.  the recall is similar but the precision has a significant reduction. Based on this result, I think it cannot improved significantly through tuning the ensemble weight.

**Comment by snukky** [comment]
Re-trained language models have higher perplexity (58 vs 46). I have trained mine with dropout (at least `--transformer-dropout 0.1`) that is missing in the published scripts. I will check if this is it or some bug in monolingual data preprocessing.

**Comment by GitHub-v7** [comment]
I added the dropout for training the lm and re-trained the project. And I got a similar perplexity (46.7534) for lm.1. Even though the result of lm is similar, the result of transformer model did not improved. Transformer.2 even got a worse result (50.61 M2) on the CoNLL-2014 test set.

**Comment by GitHub-v7** [comment]
Hi, Have you found why causes the performance difference between the released transformer model and the re-trained transformer model ?

**Comment by snukky** [comment]
Mostly because I trained the released models with larger mini-batches. The scripts are updated now. Thanks.

```
==> models/transformer.2.lmdrop/eval/test2014.eval <==
Precision   : 0.5641
Recall      : 0.4078
F_0.5       : 0.5240

==> models/transformer.4.xlbatch/eval/test2014.eval <==
Precision   : 0.6050
Recall      : 0.3998
F_0.5       : 0.5487
```
